Add option to focus the search bar when double tapping the search tab

rizwankce opened the pull request:
## Changes

- Enables double tap for tab bar(s)
- Hides double tap for me tab (as per previous comment)
- Adds search tab double tab option to focus the search bar

## Issue

closes #748 

## Screenshots

In this below GIF, double tapping the search tab opens the search bar. Single tap only opens the search screen.

![Simulator Screen Recording - iPhone 15 - 2023-11-17 at 14 48 20](https://github.com/mastodon/mastodon-ios/assets/12063704/3f3ade3b-7264-4d75-989b-26971430e0d9)


## Notes
I'm new to the combine world. Please let me know if there is anything else Im missing out or needs a different approach for this one.

rizwankce committed:
Add option to focus the search bar when double tapping the search tab bar

- Enables double tap for tab bar(s)
- Hides double tap for me tab (as per previous comment)
- Adds search tab double tab option to focus the search bar

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/mastodon/mastodon-ios?pullRequest=1169) <br/>All committers have signed the CLA.

zeitschlag commented on Mastodon/Scene/Root/MainTab/MainTabBarController.swift:
```diff
@@ -292,12 +292,11 @@ extension MainTabBarController {
-//        tabBarDoubleTapGestureRecognizer.addTarget(self, action: #selector(MainTabBarController.tabBarDoubleTapGestureRecognizerHandler(_:)))
-//        tabBarDoubleTapGestureRecognizer.delaysTouchesEnded = false
-//        tabBar.addGestureRecognizer(tabBarDoubleTapGestureRecognizer)
+        let tabBarDoubleTapGestureRecognizer = UITapGestureRecognizer()
+        tabBarDoubleTapGestureRecognizer.numberOfTapsRequired = 2
+        tabBarDoubleTapGestureRecognizer.addTarget(self, action: #selector(MainTabBarController.tabBarDoubleTapGestureRecognizerHandler(_:)))
+        tabBarDoubleTapGestureRecognizer.delaysTouchesEnded = false
+        tabBar.addGestureRecognizer(tabBarDoubleTapGestureRecognizer)
```
If you just enable it in `MainTabBarcontroller`, this works only for iPhone. Please have a look at `SidebarViewController` as that's the sidbar on iPad :)

rizwankce replied:
Oh totally forgot about the iPad support. I will make the necessary change and update you

zeitschlag commented on Mastodon/Scene/Root/MainTab/MainTabBarController.swift:
```diff
@@ -359,17 +358,22 @@ extension MainTabBarController {
-                self.coordinator.setup()
-            }
+            // double tapping search tab opens the search bar without additional taps
+            searchViewController?.searchBarTapPublisher.send("")
```
```suggestion
            searchViewController?.searchBar.becomeFirstResponder()
```

What do you think about this?

zeitschlag commented on Mastodon/Scene/Root/MainTab/MainTabBarController.swift:
```diff
@@ -359,17 +358,22 @@ extension MainTabBarController {
         switch tab {
-        case .me:
-            guard let authContext = authContext else { return }
+        // todo: reconsider the "double tap to change account" feature -> https://github.com/mastodon/mastodon-ios/issues/628
+//        case .me:
+//            guard let authContext = authContext else { return }
+//            assert(Thread.isMainThread)
+//
+//            guard let nextAccount = context.nextAccount(in: authContext) else { return }
+//            
+//            Task { @MainActor in
+//                let isActive = try await context.authenticationService.activeMastodonUser(domain: nextAccount.domain, userID: nextAccount.userID)
+//                guard isActive else { return }
+//                self.coordinator.setup()
+//            }
```
Good idea to comment this out, what do you think about removing this? We could add it as a comment to #628 so that we don't loose it, but clean up the code a bit?

rizwankce replied:
sure thing! Don't know the reason why we kept the todo so thought of not completely removing it. 

Now that you are okay with that. I will go ahead and do some clean up

zeitschlag reviewed (commented):
@rizwankce Thank you for your contribution, works for me 👍 

I added some comments, can you please have a look at them? Please consider them (mostly) feedback, i.e.: You don't have to follow them. Feel free to mark them as resolved or answer with a comment of your own :)

rizwankce committed:
Clean up commented out code related to me tab

rizwankce commented:
@zeitschlag I have made some requested changed. 

When I try to add support for iPad. Found a crash which is already happening in current version as well. So need bit more time and debugging to make it happen for iPad.

When try to focus on search in iPad landscape mode, app is crashing. The reason for the crash is search bar intrinsic size is having height as 0.0. I tried to debug more and end up figuring out `CustomSearchController` having issue(s) where it sets Search bar size by frame. 

So we can land this PR for iPhone support and once I fix the crash I will raise separate PR for iPad support.

rizwankce committed:
Add double tap top focus search bar for iPad

rizwankce committed:
Fix crash on iPad non primary display focusing search bar

Crash happens when we try to focus search bar when in iPad portrait because of custom hight search bar don't have correct intrinsic size

rizwankce commented:
@zeitschlag so after couple of debugging and trying to find root cause. I finally managed to make it work for iPad as well as fix a crash which was happening on iPad portrait mode as well. 

Now its ready for your review.

### Crash background
when focusing on search bar in iPad portrait mode, app is crashing because of custom hight / width search bar we have and its intrinsic size is 0.0. When searching for a root cause figured out that we have search controller added to navigation item.

It has new method which automatically positions the search controller w/ search bar via `navigationItem.preferredSearchBarPlacement = .stacked`

Now all works!!!

zeitschlag reviewed (approved):
LGTM! Thank you @rizwankce 🚀
